src/map/status.c

```diff
@@ -2388,8 +2388,17 @@ static int status_calc_pc_(struct map_session_data *sd, enum e_status_calc_opt o
 		calculating = 0;
 		return 0;
 	}
-	if(memcmp(b_skill,sd->status.skill,sizeof(sd->status.skill)))
+
+	if (memcmp(b_skill, sd->status.skill, sizeof(sd->status.skill))) {
+		// Client doesn't delete unavailable skills even if we refresh
+		// the skill tree, individually delete them.
+		// TODO: Check what client versions are affected
+		for (i = 0; i < MAX_SKILL_DB; i++) {
+			if (b_skill[i].id != 0 && sd->status.skill[i].id == 0)
+				clif->deleteskill(sd, b_skill[i].id);
+		}
 		clif->skillinfoblock(sd);
```

Comment:
I left it there in case older clients don't work like that, I couldn't test it with older clients.
There're additional cases like equiping an item that grants a skill or status changes like the soul linker effects.
The alternative is calling clif->addskill instead of sending the whole tree, but as I couldn't test in older clients, I wasn't sure. At least this won't break anything.

Reply:
Done. After some testing with 2012, 2013, 2014, 2019 and 2020 clients and checking this [issue](https://github.com/rathena/rathena/pull/5931]) in rAthena, the fix has been isolated to certain clients.